### DISCUSSION TITLE
Consider replacing librsvg with libresvg

### DISCUSSION THREAD

**Discussion by gunta** [libvips/libvips#2048] [Ideas]
https://github.com/RazrFalcon/resvg

Why:
- resvg substantially passes more tests than librsvg
- resvg is fast. Supports 4 rendering backends currently: cairo, skia, qt, raqote

This would help close many svg related issues.

The table of support is here: 
https://razrfalcon.github.io/resvg-test-suite/svg-support-table.html

Both library goals are different: librsvg is a GNOME library. librsvg is a cross-platform one with a multiple backend support.

**Comment by jcupitt** [comment]
Hello @gunta,

That's interesting, I hadn't heard of resvg. Sure, let's tag this as an enhancement and do some experiments.

**Comment by jcupitt** [comment]
@lovell I know you've been battling some SVG issues, you might be interested.

**Comment by lovell** [comment]
I'm rather ambivalent on this as librsvg is under heavy (re)development at the moment and is in a much better place than it was a couple of years ago when resvg was started.

It's worth noting that resvg:

* is removing support for "backends" other than raqote, which is not a bad thing, especially as librsvg is considering likewise - see https://github.com/RazrFalcon/resvg/issues/195 and https://gitlab.gnome.org/GNOME/librsvg/issues/530
* is not yet widely available via package manager - Debian unstable added it a year ago but hasn't updated recently

**Comment by gunta** [comment]
>> is removing support for "backends" other than raqote, which is not a bad thing, especially as librsvg is considering likewise

> According to my "benchmarks", raqote isn't that slow, but still considerably slower than Skia. I hope that this can be fixed soon. I do not plan to remove other backends right now.

So the change its not happening yet, but definitely Raqote, being used in Servo has good potential.

> is not yet widely available via package manager - Debian unstable added it a year ago but hasn't updated recently

Since neither library seems to aim for 100% full support, 
there will be always a [mismatch with the spec](https://razrfalcon.github.io/resvg-test-suite/svg-support-table.html), and users of libvips might need one or another depending on what the `svg` content looks like or performance-wise.

Since we will not have a clear winner anytime soon, I suggest to support both `librsvg` and `resvg` as a compile-time option.

In my case I need to use the `image` SVG element `External PNG` feature of the [spec](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/image) which `resvg` does support.

# resvg goals
> resvg can be used to render SVG files based on a static SVG Full 1.1 subset, excluding fonts support. 
In simple terms: no animations, scripting and fonts.
The core idea is to make a fast, portable, small, multiple backend library.

# librsvg non-goals
> Implement every single SVG feature that is in the spec.
Implement scripting or external access to the SVG's DOM.
Implement support for CSS-based animations
Replace the industrial-strength SVG rendering machinery in modern web browsers.

**Comment by jcupitt** [comment]
Ah, I hadn't realized it was not yet widely packaged, thanks Lovell. I think without an easy way to test and deploy, any new libvips code is likely to bitrot.

Let's look at this again when resvg is packaged.

**Comment by kleisauke** [comment]
Here's a interesting RFC on Wikimedia regarding librsvg/resvg: https://phabricator.wikimedia.org/T40010.

For wasm-vips, a resvg loader might also be useful, since librsvg is currently difficult to compile for WebAssembly because it requires many dependencies (see https://github.com/kleisauke/wasm-vips/issues/4#issuecomment-960969507). This is not a issue on the web, since you can use a canvas to render an SVG (live demo [here](https://kleisauke.github.io/wasm-vips/playground/?deflate=tVRhb9s2EP0rhDtAMupQitOsgeMUWLJlCNB2Q5MMA-oCpSXKZiORKkVbdoP8974TZTtykbZfahg2j7x79_juju97UcTmzpXVKIoqJ5I7s5Q2y03NE1NEIjo6Gg5fxvHxRGcLnThlNEutqK90rrS8_u_vsFrO_splIbUbsMSt8CPyfAqgPrufaIZPYnTlGPxu371mZ0zLmv3_5vW1tErk6ou0YZ9XG-PGXDur9OwRbH_gYeijilmLcFWImQz7pxPtT3HCjc6NSOGwoRpuOTQ8hF6KitcqdXM4UUSzBsaey1yq2dy1Pt547ORWvNGgoeDmqhqwGF8iswPyIjQMaeNhR7OyCaCDVDgxUoQR4bLPV0V-ypK5sJV0ZwuXnQxYwJ4zqROTytt3VxemKI2GHiQNlCRggE5wTvJ65sBNTbIg2XhipXCyFTEMvEOwUROXgHN735l0F0Y7uYLfMA28qt0yb2E_L6RdX8tcJs7YMAAZYPrKQ-6zV92q475_4p6UCqohTyMabYUkGXXLrihby0veLTwglqqseAPA0QKX1hRvZGHsOmyzcJL0u5AD9mLgYc6FTi-NLYTjC1J910kYiEulUb_1gNVWOcncXDIrq0XumDNMsGlupo8v6WKQKzE2hKcTsDP1tvDexSzc-SLLpG3b91Zpd_KHtYLI8ybNjfEeYcBLDVG7AO7wiRw7J5NLnptZ-PEC5IlpBxYb5o79dg-gAzB-YIXKc1VJRKYV_9iFcs0dW67nWIbvtzf4MGD3zK1LOWKBb1-iyx66fCnezzt-2178Z_oJXQM7pNOuv5_sp5oXp01XdkeozdHub2PRZW3g-foqDQMwLxcu6HNRllKnF3OVp9QyzQT5e3sOpZXf4YDT7fS4Fb0Oe543mJ-3GNawaTBAZmoWNmVEaCc5wmn7ScY-do8xQCioN-iNsXyFoRxHtJjoMYaQ4f3Q1dmkR4853vK6rnl9xI2dRcM4jumJmfRYMxNwOo5jWH4mNiaAGBtbVIitsHdCHmssfqfFJnB43An0piX_Ia0qt84ljAy9NXqWZcgcn2LXmjs5ekbWzj5oMEfDcnU66UV72Q9fbNIfNsA_zt8EdPNTtiz7ifyM_A9MKRLl1qOYv_x1jEiTLqMk2WN0_A2jQx63jMZUSPqnyv_7dtcCqVoylSKR73aUcxxhb-Oaqyl1JQ0bOuvbKL__KApt1vvwFQ)), but for Node.js this won't work.

**Comment by barracuda156** [comment]
@gunta Unfortunately, looks like this new library is again depending on Rust.

**Comment by JoKalliauer** [comment]
In 2021 I did a study for Wikipedia: https://commons.wikimedia.org/wiki/User:JoKalliauer/SVG_test_suites

And for the SVGs on Wikipedia resvg is faster and more correct than rsvg, expecially the bugs in rsvg were imho more sevier.

**Comment by gemini133** [comment]
@jcupitt 
Hi John, will resvg be an offical loader of libvips? been using @kleisauke fork of resvg support, so far so good.
